atlasdb-impl-shared/src/test/java/com/palantir/atlasdb/transaction/impl/expectations/KeyValueServiceDataTrackerTest.java

```diff
@@ -0,0 +1,306 @@
+        tracker.recordReadForTable(TABLE_1, KVS_METHOD_NAME_1, BYTES_READ_1);
+        tracker.recordReadForTable(TABLE_1, KVS_METHOD_NAME_2, BYTES_READ_3);
+        tracker.recordReadForTable(TABLE_1, KVS_METHOD_NAME_3, BYTES_READ_3);
+        Consumer<Long> tableOneFirstTracker = tracker.recordCallForTable(TABLE_1);
+        tableOneFirstTracker.accept(BYTES_READ_3);
+        Consumer<Long> tableOneSecondTracker = tracker.recordCallForTable(TABLE_1);
+        tableOneSecondTracker.accept(BYTES_READ_2);
+
+        tracker.recordReadForTable(TABLE_2, KVS_METHOD_NAME_2, BYTES_READ_2);
+        tracker.recordReadForTable(TABLE_2, KVS_METHOD_NAME_3, BYTES_READ_3);
+        Consumer<Long> tableTwoTracker = tracker.recordCallForTable(TABLE_2);
+        tableTwoTracker.accept(BYTES_READ_1);
+        tableTwoTracker.accept(BYTES_READ_2);
+
+        tracker.recordReadForTable(TABLE_3, KVS_METHOD_NAME_1, BYTES_READ_2);
+        Consumer<Long> tableThreeTracker = tracker.recordCallForTable(TABLE_3);
+        tableThreeTracker.accept(BYTES_READ_1);
+        tableThreeTracker.accept(BYTES_READ_1);
+        tableThreeTracker.accept(BYTES_READ_2);
+
+        tracker.recordTableAgnosticRead(KVS_METHOD_NAME_4, BYTES_READ_1);
+        tracker.recordTableAgnosticRead(KVS_METHOD_NAME_5, BYTES_READ_2);
+
+        assertThat(tracker.getReadInfo())
+                .isEqualTo(createTransactionReadInfo(
+                        5 * BYTES_READ_1 + 6 * BYTES_READ_2 + 4 * BYTES_READ_3,
+                        12,
+                        createMaximumBytesKvsCallInfo(BYTES_READ_3, KVS_METHOD_NAME_2)));
+
+        assertThat(tracker.getReadInfoByTable())
+                .containsExactlyEntriesOf(ImmutableMap.of(
+                        TABLE_1,
+                                createTransactionReadInfo(
+                                        BYTES_READ_1 + BYTES_READ_2 + 3 * BYTES_READ_3,
+                                        5,
+                                        createMaximumBytesKvsCallInfo(BYTES_READ_3, KVS_METHOD_NAME_2)),
+                        TABLE_2,
+                                createTransactionReadInfo(
+                                        BYTES_READ_1 + 2 * BYTES_READ_2 + BYTES_READ_3,
+                                        3,
+                                        createMaximumBytesKvsCallInfo(BYTES_READ_3, KVS_METHOD_NAME_3)),
+                        TABLE_3,
+                                createTransactionReadInfo(
+                                        2 * BYTES_READ_1 + 2 * BYTES_READ_2,
+                                        2,
+                                        createMaximumBytesKvsCallInfo(BYTES_READ_2, KVS_METHOD_NAME_1))));
+    }
+
+    /**
+     * Schedules tracking calls from {@link #differentMethodCallsAreTracked} repeatedly ({@link #CONCURRENCY_ROUNDS}
+     * times.
+     */
+    @Test
+    public void concurrencyAndOrderOfExecutionTest() throws InterruptedException {
```

Comment:
I tried to force a concurrent modification exception by adding some dummy hash maps to `KeyValueServiceDataTracker` but I couldn't. I think the test is fine though